node/core/approval-voting/src/criteria.rs

```diff
@@ -483,50 +619,100 @@ pub(crate) fn check_assignment_cert(
 	let public = schnorrkel::PublicKey::from_bytes(validator_public.as_slice())
 		.map_err(|_| InvalidAssignment(Reason::InvalidAssignmentKey))?;
 
-	if claimed_core_index.0 >= config.n_cores {
-		return Err(InvalidAssignment(Reason::CoreIndexOutOfBounds))
+	// Check that we have all backing groups for claimed cores.
+	if claimed_core_indices.is_empty() && claimed_core_indices.len() != backing_groups.len() {
+		return Err(InvalidAssignment(Reason::InvalidArguments))
 	}
 
 	// Check that the validator was not part of the backing group
 	// and not already assigned.
-	let is_in_backing =
-		is_in_backing_group(&config.validator_groups, validator_index, backing_group);
+	for (claimed_core, backing_group) in claimed_core_indices.iter().zip(backing_groups.iter()) {
+		if claimed_core.0 >= config.n_cores {
+			return Err(InvalidAssignment(Reason::CoreIndexOutOfBounds))
+		}
 
-	if is_in_backing {
-		return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		let is_in_backing =
+			is_in_backing_group(&config.validator_groups, validator_index, *backing_group);
+
+		if is_in_backing {
+			return Err(InvalidAssignment(Reason::IsInBackingGroup))
+		}
 	}
 
 	let &(ref vrf_output, ref vrf_proof) = &assignment.vrf;
-	match assignment.kind {
-		AssignmentCertKind::RelayVRFModulo { sample } => {
-			if sample >= config.relay_vrf_modulo_samples {
+	match &assignment.kind {
+		AssignmentCertKindV2::RelayVRFModuloCompact { sample, core_indices } => {
+			if *sample >= config.relay_vrf_modulo_samples {
 				return Err(InvalidAssignment(Reason::SampleOutOfBounds))
 			}
 
 			let (vrf_in_out, _) = public
 				.vrf_verify_extra(
-					relay_vrf_modulo_transcript(relay_vrf_story, sample),
+					relay_vrf_modulo_transcript(relay_vrf_story, *sample),
 					&vrf_output.0,
 					&vrf_proof.0,
-					assigned_core_transcript(claimed_core_index),
+					assigned_cores_transcript(core_indices),
 				)
 				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
```

Comment:
We do not check `core_indices` agrees with the results from `vrf_in_out` anywhere?

Reply:
As noted below, we could just check that core_indices is a subset of those determined by the vrf.

Reply:
It's checked a few lines below.